File: BIT.py
class BIT:
    def __init__(self, size: int):
        self.size = size
        self.tree = [0] * (size + 1)

    def update(self, index: int, delta: int):
        idx = index + 1
        while idx <= self.size:
            self.tree[idx] += delta
            idx += idx & -idx

    def query(self, index: int) -> int:
        idx = index + 1
        s = 0
        while idx > 0:
            s += self.tree[idx]
            idx -= idx & -idx
        return s

    def query_range(self, left: int, right: int) -> int:
        if left > right:
            return 0
        sum_right = self.query(right)
        sum_left_minus_1 = self.query(left - 1) if left > 0 else 0
        return sum_right - sum_left_minus_1

File: BIT.cpp
template <typename T>
struct BIT {
    int n;
    std::vector<T> tree;

    // Initialize with size n
    BIT(int n) : n(n), tree(n + 1, 0) {}

    // Point update: Add 'delta' to the value at 0-based index 'i'
    void add(int i, T delta) {
        for (++i; i <= n; i += i & -i) tree[i] += delta;
    }

    // Prefix sum: Returns the sum in range [0, i)
    T query(int i) const {
        T sum = 0;
        for (; i > 0; i -= i & -i) sum += tree[i];
        return sum;
    }

    // Range sum: Returns the sum in range [l, r)
    T query(int l, int r) const {
        return query(r) - query(l);
    }
};
